---
title: "Create data package"
author: "Trang Tran"
date: "March 22, 2020"
output: html_notebook
---

```{r setup, include=FALSE, echo=FALSE}
library(magrittr)
library(ggplot2)
library(Biobase)

options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = TRUE)

file.metadata_sra = 'PRJNA432208_metadata_cleaned.tsv'
file.metadata_fastq = 'fastq_metadata.tsv'
```

## Overview

## Download count matrices and meta data, set environment variable 'DBDIR'

```{r,eval = FALSE}
Sys.setenv('DBDIR' = './')
```


## Gene-level `ExpressionSet`

### Assay data

```{r}
gene.counts = readRDS(file.path(Sys.getenv('DBDIR'), 'matrix.gene.expected_count.RDS'))
gene.tpm = readRDS(file.path(Sys.getenv('DBDIR'), 'matrix.gene.tpm.RDS'))
gene.assayData <- new.env()
assign('exprs', gene.counts, envir = gene.assayData) # exprs is required
assign('expected_count', gene.counts, envir = gene.assayData)
assign('tpm', gene.tpm, envir = gene.assayData)
ExpressionSet(assayData = gene.assayData)
```


### Phenotypic data

```{r}
meta1 = read.table(file.path(Sys.getenv('DBDIR'), file.metadata_sra), sep = '\t', header = TRUE)
meta2 = read.table(file.path(Sys.getenv('DBDIR'), file.metadata_fastq), sep = '\t', header = TRUE)
```

```{r}
starLog = read.table(file.path(Sys.getenv('DBDIR'),
                               'starLog.final.tsv'), sep = '\t',header=TRUE)
meta3 = data.frame(t(starLog[,-c(1,2)])) %>%
    set_colnames(paste0('starAlignment.',starLog[[1]])) %>%
    set_rownames(colnames(starLog)[-c(1,2)])
meta3$RunId = rownames(meta3)
meta3$BiosampleId <- meta3$RunId
meta3$RunId <- NULL
str(meta3)
```

```{r}
libnames <- colnames(gene.assayData$exprs) %>% as.character()
allMeta <- data.frame('BiosampleId' = libnames) %>%
    plyr::join(y = meta1, by = 'BiosampleId', type = 'left')  %>%
    plyr::join(y = meta2, by = 'RunId', type = 'left')  %>%
    plyr::join(y = meta3, by = 'BiosampleId', type = 'left') %>%
    set_rownames(libnames)
colDescription = colnames(allMeta)
colDescription[grepl('starAlignment', colnames(allMeta))] = starLog[[2]]
phenoData <-
    Biobase::AnnotatedDataFrame(data = allMeta, varMetadata = data.frame('labelDescription' = colDescription, row.names = colnames(allMeta)))
phenoData@varMetadata
```

Making sure sample order is preserved

```{r}
all(phenoData@data$BiosampleId == colnames(gene.counts))
```

### Annotations and features

```{r}
gene.featureData <- data.frame('ID' = rownames(gene.assayData$exprs))
rownames(gene.featureData) = gene.featureData$ID
gene.featureData = gene.featureData %>%
    Biobase::AnnotatedDataFrame(data = ., varMetadata = data.frame('labelDescription' = colnames(.), row.names = colnames(.)))
gene.featureData
```

### Experiment description

```{r}
experimentData <- new("MIAME",
                      lab = "PCRM-NICHD",
                      contact = "arnaouta@mail.nih.gov",
                      title = "Gene expression profiling of IRBIT mutant flies",
                      abstract = "The maintenance of the intestinal epithelium is ensured by the controlled proliferation of intestinal stem cells (ISCs) and differentiation of their progeny into various cell types, including enterocytes (ECs) that both mediate nutrient absorption and provide a barrier against pathogens. The signals that regulate transition of proliferative ISCs into differentiated ECs are not fully understood. IRBIT is an evolutionarily conserved protein that regulates ribonucleotide reductase (RNR), an enzyme critical for the generation of DNA precursors. Here, we show that IRBIT expression in ISC progeny within the Drosophila midgut epithelium cells regulates their differentiation via suppression of RNR activity. Disruption of this IRBIT-RNR regulatory circuit causes a premature loss of intestinal tissue integrity. Furthermore, age-related dysplasia can be reversed by suppression of RNR activity in ISC progeny. Collectively, our findings demonstrate a role of the IRBIT-RNR pathway in gut homeostasis.",
                      other = list(
                          citation = "Arnaoutov A, Lee H, Plevock Haase K, Aksenova V et al. IRBIT Directs Differentiation of Intestinal Stem Cell Progeny to Maintain Tissue Homeostasis. iScience 2020 Mar 27;23(3):100954. PMID: 32179478"
                      )
                      )
```

### Assembling an `ExpressionSet`

```{r}
dmelirbit.rnaseq.gene <- ExpressionSet(assayData = gene.assayData,
                             phenoData = phenoData,
                             experimentData = experimentData,
                             featureData = gene.featureData,
                             annotation = 'Ensembl Dmel_BDGP6.28 + ERCC-92'
                             )
```

## Transcript-level `ExpressionSet`

### Assay data

```{r}
transcript.counts = readRDS(file.path(Sys.getenv('DBDIR'), 'matrix.transcripts.expected_count.RDS'))
transcript.tpm = readRDS(file.path(Sys.getenv('DBDIR'), 'matrix.transcripts.tpm.RDS'))
transcript.assayData <- new.env()
assign('exprs', transcript.counts, envir = transcript.assayData) # exprs is required
assign('expected_count', transcript.counts, envir = transcript.assayData)
assign('tpm', transcript.tpm, envir = transcript.assayData)
ExpressionSet(assayData = transcript.assayData)
```

### Phenotypic data

Same as above

### Annotations and features

```{r}
feature_attrs = read.table(file.path(Sys.getenv('DBDIR'), 'feature_attrs.rsem.transcripts.tsv'), sep = '\t', header=TRUE, colClasses = 'character')
transcript.featureData <- data.frame('transcript_id' = rownames(transcript.assayData$exprs)) %>%
    plyr::join(y = feature_attrs[, c('transcript_id', 'gene_id', 'length', 'effective_length')], by = 'transcript_id', type = 'left') %>%
    set_names(c('ID', 'GeneId', 'Length', 'EffectiveLength'))
rownames(transcript.featureData) = transcript.featureData$ID
transcript.featureData = transcript.featureData %>%
    Biobase::AnnotatedDataFrame(data = ., varMetadata = data.frame('labelDescription' = colnames(.), row.names = colnames(.)))
```

### Experiment description

Same as above

### Assembling an `ExpressionSet`

```{r}
dmelirbit.rnaseq.transcript <- ExpressionSet(assayData = transcript.assayData,
                             phenoData = phenoData,
                             experimentData = experimentData,
                             featureData = transcript.featureData,
                             annotation = 'Ensembl Dmel_BDGP6.28 + ERCC-92'
                             )
```

## Writing data sets

```{r,eval=FALSE}
usethis::use_data(dmelirbit.rnaseq.gene,internal = FALSE, overwrite = TRUE)
usethis::use_data(dmelirbit.rnaseq.transcript,internal = FALSE, overwrite = TRUE)
```

